rust/osm2lanes/src/transform/tags_to_lanes/mod.rs

```diff
@@ -566,58 +568,75 @@ pub fn tags_to_lanes(
-        };
-        half - both_ways
-    } else if tags.is("lanes:bus", "2") {
-        2
-    } else {
-        1
-    };
-    let num_driving_back = if let Some(n) = tags
+        .and_then(|num| num.parse::<usize>().ok());
+    let tagged_backward = tags
         .get("lanes:backward")
-        .and_then(|num| num.parse::<usize>().ok())
-    {
-        n
-    } else if let Some(n) = tags.get("lanes").and_then(|num| num.parse::<usize>().ok()) {
-        let base = n - num_driving_fwd;
-        let half = if oneway.into() {
-            base
-        } else {
-            // lanes=1 but not oneway... what is this supposed to mean?
-            base.max(1)
-        };
-        half - both_ways
-    } else if tags.is("lanes:bus", "2") {
-        if oneway.into() {
-            1
-        } else {
-            2
+        .and_then(|num| num.parse::<usize>().ok());
+    let tagged_both_ways = tags
+        .get("lanes:both_ways")
+        .and_then(|num| num.parse::<usize>().ok());
+    let num_both_ways = tagged_both_ways.unwrap_or(0);
+    let tagged_bus_lanes = tags
+        .get("lanes:bus")
+        .and_then(|num| num.parse::<usize>().ok());
+
+    if oneway.into() {
+        // TODO support oneway=-1
+        if tagged_backward.is_some() {
+            warn!("lanes:backwards tag on a oneway road")
+        }
+        if tagged_both_ways.is_some() {
+            warn!("lanes:both_ways tag on a oneway road")
+        }
+        match (tagged_lanes, tagged_forward) {
+            (Some(l), None) => (l, 0),
+            (Some(l), Some(f)) => {
+                if l != f {
+                    warn!("lanes tag does not agree with lanes:forward on oneway road")
+                }
+                (f, 0)
+            }
+            // Without the "lanes" tag, assume lanes:bus adds onto the assumed single lane
+            (None, _) => (
+                tagged_forward.unwrap_or(1) + tagged_bus_lanes.unwrap_or(0),
+                0,
+            ),
         }
-    } else if oneway.into() {
-        0
     } else {
-        1
-    };
-    (num_driving_fwd, num_driving_back)
+        match (tagged_lanes, tagged_forward, tagged_backward) {
+            (_, Some(f), Some(b)) => {
+                if tagged_lanes.is_some_and(|&l| l != f + b + num_both_ways) {
+                    warn!("lanes tag does not agree with lanes:forward, lanes:backward and lanes:both_ways");
+                }
+                (f, b)
+            }
+            (Some(l), Some(f), None) => (f, l - f - num_both_ways),
+            (Some(l), None, Some(b)) => (l - b - num_both_ways, b),
+            (Some(1), None, None) => {
+                warn!("lanes=1 on a two way road");
+                (1, 1) // assume they meant lanes=2, I guess
+            }
+            (Some(l), None, None) => {
+                // Try to divide the lanes equally.
+                if (l - num_both_ways) % 2 != 0 {
+                    warn!("lanes - lanes:both_ways cannot be evenly distributed");
```

Comment:
hmmmmmmm
@dabreegster I think you made the decision to round in the favour of forward direction.
Do you think that is consistent enough to not raise a warning?

Reply:
A warning sounds appropriate. Unless the OSM wiki is very clear what this case means, it's a case of missing / ambiguous data. Good catch!

Reply:
https://wiki.openstreetmap.org/wiki/Key:lanes#Extended_tagging

I agree with the warning, no change needed.